ot/bregman.py

```diff
@@ -1787,3 +1789,287 @@ def empirical_sinkhorn_divergence(X_s, X_t, reg, a=None, b=None, metric='sqeucli
+    # ----------------------------------------------------------------------------------------------------------------#
+
+    if ns_budget == ns and nt_budget == nt:
+        # full number of budget points (ns, nt) = (ns_budget, nt_budget)
+        I = list(range(ns))
+        J = list(range(nt))
+        epsilon = 0.0  
+        kappa = 1.0  
+
+        cst_u = 0.
+        cst_v = 0.
+
+        bounds_u = [(0.0, np.inf)] * ns
+        bounds_v = [(0.0, np.inf)] * nt
+        
+        a_I = a
+        b_J = b
+        K_min = K.min()
+        K_IJ = K
+        K_IJc = []
+        K_IcJ = []
+
+        vec_eps_IJc = np.zeros(nt)
+        vec_eps_IcJ = np.zeros(ns)
+
+    else:
+        # sum of rows and columns of K
+        K_sum_cols = K.sum(axis=1)
+        K_sum_rows = K.sum(axis=0)
+
+        if uniform:
+            if ns / ns_budget < 4:
+                aK_sort = np.sort(K_sum_cols)
+                epsilon_u_square = a[0] / aK_sort[ns_budget - 1]
+            else:
+                aK_sort = bottleneck.partition(K_sum_cols, ns_budget - 1)[ns_budget - 1]
+                epsilon_u_square = a[0] / aK_sort
+
+            if nt / ns_budget < 4:
+                bK_sort = np.sort(K_sum_rows)
+                epsilon_v_square = b[0] / bK_sort[ns_budget - 1]
+            else:
+                bK_sort = bottleneck.partition(K_sum_rows, nt_budget - 1)[nt_budget - 1]
+                epsilon_v_square = b[0] / bK_sort
+        else:
+            aK = a / K_sum_cols
+            bK = b / K_sum_rows
+
+            aK_sort = np.sort(aK)[::-1]
+            epsilon_u_square = aK_sort[ns_budget - 1]
+
+            bK_sort = np.sort(bK)[::-1]
+            epsilon_v_square = bK_sort[ns_budget - 1]
+
+        # active sets I and J (see Proposition .. in [1])
+        I = np.where(a >= epsilon_u_square * K_sum_cols)[0].tolist()
+        J = np.where(b >= epsilon_v_square * K_sum_rows)[0].tolist()
+
+        if len(I) != ns_budget:
+            if uniform:
+                aK = a / K_sum_cols
+                aK_sort = np.sort(aK)[::-1]
+            epsilon_u_square = aK_sort[ns_budget - 1:ns_budget + 1].mean()
+            I = np.where(a >= epsilon_u_square * K_sum_cols)[0].tolist()
+
+        if len(J) != nt_budget:
+            if uniform:
+                bK = b / K_sum_rows
+                bK_sort = np.sort(bK)[::-1]
+            epsilon_v_square = bK_sort[ns_budget - 1:ns_budget + 1].mean()
+            J = np.where(b >= epsilon_v_square * K_sum_rows)[0].tolist()
+
+        epsilon = (epsilon_u_square * epsilon_v_square) ** (1 / 4)
+        kappa = (epsilon_v_square / epsilon_u_square) ** (1 / 2)
+
+        if verbose:
+            print("Epsilon = %s\n" %epsilon)
+            print("Scaling factor = %s\n" %kappa)
+
+        if verbose:
+            print('|I_active| = %s \t |J_active| = %s ' %(len(I), len(J)))
+
+        # Ic, Jc: complementary of the active sets I and J
+        Ic = list(set(list(range(ns))) - set(I))
```

Comment:
numpy.delete on an numpy.arange should also work. you should test to see which one is the more efficient.

Reply:
binary indexing is better indeed and it is much easier do do the complementary.

Reply:
Actually, I compared three ways to define Ic as follows:
```
fulllist_s = list(np.arange(ns))
Ic = list(np.delete(fulllist_s, [fulllist_s.index(v) for v in I]))
Ic = list(np.array(fulllist_s)[~np.isin(fulllist_s, I)])
Ic = list(set(fulllist_s) - set(I))
```
The definition using `set` is much faster than ways 1 and 2. So, I suggest to keep it.

Reply:
No, it's the conversion array to list to array that slows it down, if we're going for numpy, we must not use lists anywhere:
```python
fulllist_s = np.arange(ns)
Ic = fulllist_s[~np.isin(fulllist_s, I)]
```
 3x faster with numpy

<img width="731" alt="Screenshot 2020-01-07 at 19 19 06" src="https://user-images.githubusercontent.com/17051500/71918461-998ebf00-3182-11ea-9d84-74c67973b02e.png">

Reply:
yes binary indexing is faster but it needs to stay in full numpy. From the algorithm it seems to me it is feasible and the code might even be easier to understand with binary indexing.